butane_core/src/autopk.rs

```diff
@@ -0,0 +1,118 @@
+//! Contains the [AutoPk] type for autoincrementing primary keys.
+
+use super::{FieldType, FromSql, PrimaryKeyType, Result, SqlType, SqlVal, SqlValRef, ToSql};
+use serde::{Deserialize, Serialize};
+use std::cmp::{Ordering, PartialOrd};
+
+/// Wrapper around a [PrimaryKeyType] to indicate the the primary key
+/// will be initialized automatically when the object is created in
+/// the database.
+#[derive(Clone, Debug, Default, Serialize, Deserialize)]
+pub struct AutoPk<T: PrimaryKeyType> {
+    inner: Option<T>,
+}
+
+impl<T: PrimaryKeyType> AutoPk<T> {
+    /// Create an uninitialized value for an object which has not yet been saved.
+    pub fn uninitialized() -> Self
+    where
+        T: Default,
+    {
+        Self::default()
+    }
+
+    /// Create an initialized primary key value for a previously saved
+    /// object. You do not usually need to call this directly (it will
+    /// happen implicitly when you load from the database).
+    fn with_value(val: T) -> Self {
+        AutoPk { inner: Some(val) }
+    }
+}
+
+impl<T: PrimaryKeyType> FromSql for AutoPk<T> {
+    /// Used to convert a SqlValRef into another type.
+    fn from_sql_ref(val: SqlValRef<'_>) -> Result<Self>
+    where
+        Self: Sized,
+    {
+        Ok(AutoPk::with_value(T::from_sql_ref(val)?))
+    }
+
+    /// Used to convert a SqlVal into another type. The default
+    /// implementation calls `Self::from_sql_ref(val.as_ref())`, which
+    /// may be inefficient. This method is chiefly used only for
+    /// primary keys: a more efficient implementation is unlikely to
+    /// provide benefits for types not used as primary keys.
+    fn from_sql(val: SqlVal) -> Result<Self>
+    where
+        Self: Sized,
+    {
+        Ok(AutoPk::with_value(T::from_sql(val)?))
+    }
+}
+
+impl<T: PrimaryKeyType> ToSql for AutoPk<T> {
+    fn to_sql(&self) -> SqlVal {
+        self.inner
+            .as_ref()
+            .expect("PK is not generated yet!")
```

Comment:
DRY this message? create a private func for 
```rust
.as_ref().expect("PK is not generated yet!")
```

Reply:
Updated